### ISSUE TITLE
dac loss would be nan/inf after some train steps

### ISSUE THREAD

**Issue by Pydataman** [youngsheen/SimVQ#6]
audio sr：16000
other params not change

**Comment by youngsheen** [comment]
I have not tested the model at a 16 kHz sample rate. The current configuration is optimized for 24 kHz. To support 16 kHz, you may need to adjust the model's hyperparameters accordingly, including the downsampling rate, discriminator settings, and other relevant factors.

**Comment by HaiFengZeng** [comment]
same error+1, change to 16khz and adjust downsampling rate to [8,5,4,4], and get nan after 20k train steps in mixed language dataset, but can success in train libritts

**Comment by erogol** [comment]
+1 with 24khz dataset

**Comment by youngsheen** [comment]
Could you please let me know which dataset you are using? I may be able to run an experiment to investigate the issue. Additionally, have you tried using the same settings with the WavTokenizer repository?

**Comment by kunsung** [comment]
same error +1，and I encountered the same error while using the quantization method in my own model framework.

**Comment by kunsung** [comment]
> same error +1，and I encountered the same error while using the quantization method in my own model framework.

note：trained in mandarin dataset

**Comment by youngsheen** [comment]
@kunsung Hi, could you please let me know which dataset you are using? I need more information to reproduce the issue. In my opinion, the discriminator module might be responsible for the NaN problem.

**Comment by erogol** [comment]
I figured gradient clipping helps.

**Comment by youngsheen** [comment]
@erogol Glad to see the success! The gradient clipping seems to indicate instability in the discriminator module to me.

**Comment by erogol** [comment]
with gradient clipping, it took longer but val DAC disc losses and grad_norm of the generator layers started to go up.  Still training but I think it's not gonna work

**Comment by youngsheen** [comment]
![image](https://github.com/user-attachments/assets/a9c58277-9a56-421f-922c-43ae8c9bae11)
![image](https://github.com/user-attachments/assets/1da43f0c-399d-45f9-954a-f8398ec311ea)
Hi, this is the tensorboard log of my experiment with a codebook size of 65k for your reference.

**Comment by kunsung** [comment]
> @kunsung Hi, could you please let me know which dataset you are using? I need more information to reproduce the issue. In my opinion, the discriminator module might be responsible for the NaN problem.

I used a private dataset, which is a mixture of Chinese and English data and tens of thousands of hours in size. I think the problem is not with the discriminator, because the gradient explosion also occurs when I only use the generator with reconstruction loss. I have tried using learning rates of different sizes, but none of them worked. At the same time, I tried two steps training, first without quantization, reconstructing the continuous vector, and then adding quantization training, but it also did not solve the problem. I think the effectiveness of the method may still need to be verified on large-scale data.

**Comment by youngsheen** [comment]
> > @kunsung Hi, could you please let me know which dataset you are using? I need more information to reproduce the issue. In my opinion, the discriminator module might be responsible for the NaN problem.
> 
> I used a private dataset, which is a mixture of Chinese and English data and tens of thousands of hours in size. I think the problem is not with the discriminator, because the gradient explosion also occurs when I only use the generator with reconstruction loss. I have tried using learning rates of different sizes, but none of them worked. At the same time, I tried two steps training, first without quantization, reconstructing the continuous vector, and then adding quantization training, but it also did not solve the problem. I think the effectiveness of the method may still need to be verified on large-scale data.

I succeeded all experiments with various codebook size on imagenet (large enough I think) and libritts (I provide the log above). I am open and glad to help figure out the issues, but the private dataset you claimed and the model framework you implemented cannot be reproduced by anyone. If you don't provide us with a reproducible environment, there is nothing we can do to assist you in this setting. 
I recommend starting your tests using our code and dataset without making any modifications and check the results.

**Comment by erogol** [comment]
Thanks for sharing the loss curves. I'd say it is similar to my curves but it is still interesting to see the mod collapse of the discriminators even in your logs. Do you know why it happens and doesn't interfere with the generator?

**Comment by erogol** [comment]
it maybe about the order of updates though. In general, you need to update the discriminator first but in your code it is reversed.

**Comment by soutlu** [comment]
This issue also appears in Open-MAGVIT2. When I train on my own dataset for about 20 epochs or 400,000 steps, the model crashes and all the weights become NaN.So this should be an issue with Open-MAGVIT2.

**Comment by ChuxiJ** [comment]
I meet the same issue, how to make the commit loss stable?

![image](https://github.com/user-attachments/assets/8fe11826-c4b0-4686-a3b4-101fd7481393)

**Comment by youngsheen** [comment]
Here is my training log of commit_loss. What's the value of commit_weight in your experiment?
<img width="935" alt="image" src="https://github.com/user-attachments/assets/de03709b-b82c-4657-b247-9a0b38415651">

**Comment by ChuxiJ** [comment]
1000.0

normally it is the same loss level
but occasionally it gets a big loss like that 

Should the normalization module be necessary in the encoder output layer and before the quantizer?

**Comment by youngsheen** [comment]
Hi, all. When I implemented the streaming version of SimVQ on audio data, I encountered the Nan issue as well, which seems to occur randomly. I’m still investigating the cause.

I remove the attention layer in the wavtokenizer architecture, since it does not make sense to me for the "semantic" purpose. So far, the NaN issue has not reappeared. The codebook utilization converges a little bit slower, but it can still achieve 100%. 

I’ll share more details about the solution as I make further progress.
